### ISSUE TITLE
n_groups in pipeline

### ISSUE THREAD

**Issue by maxecharles** [LouisDesdoigts/amigo#37]
Hi @LouisDesdoigts you mentioned that the feature of truncating the ramp broke in the whole overhaul and I noticed there's a bunch of remnants of that code hanging around -- I actually think this is not necessary anymore now that the `jwst` pipeline is not a dependancy. I've been using this function to truncate the ramps/slopes and relevant variances after you already load them into an `Exposure`.

```python
def clip_ramp(exposures: list, n_groups: int):
    new_exposures = []

    for exp in exposures:
        up_to = 45 - n_groups
        for attr in ["slopes", "variance", "ramp", "ramp_variance"]:
            exp = exp.set(attr, exp.get(attr)[:-up_to, ...])
        new_exposures.append(exp)

    return new_exposures
```

So I guess I'm saying, if you wanna get rid of it entirely from the pipeline that's definitely fine - unless you see something wrong with this way of doing things.

**Comment by maxecharles** [comment]
What I've been doing for the last year or so is using the function [`dorito.misc.truncate_files`](https://github.com/maxecharles/dorito/blob/5a0b808370c7bb69134aeb97ac678bacd38af762/src/dorito/misc.py#L4), where you just call it on all your files after you load them in. It should handle the uncertainties too. This has worked well for me.

Perhaps this function actually belongs in `amigo`?

```python
def truncate_files(files, ngroups):
    """
    Truncate the ramp of files to only have ngroups. This is use for saturated or highly nonlinear data where you want to discard the top of the ramp.

    Args:
        files (list): List of calslope files to truncate.
        ngroups (int): Number of groups to keep in the ramp.
    """

    for file in files:

        # grabbing the total number of groups in the ramp
        top_group = file["RAMP"].data.shape[0]
        up_to = top_group - ngroups

        # files are mutable, so they will change in place
        for attr in ["RAMP", "SLOPE", "RAMP_SUP", "SLOPE_SUP"]:
            file[attr].data = file[attr].data[:-up_to, ...]
        for attr in ["RAMP_COV", "SLOPE_COV"]:
            file[attr].data = file[attr].data[:-up_to, :-up_to, ...]
```

**Comment by LouisDesdoigts** [comment]
I think we should put something like this Amigo, but I also want to (and will need to) develop some form of dynamic pixel masking once we learn the limits of the EDM around saturation levels. 

I think this is relative achievable, just super annoying, by saving the pixels as a list/tuple/dict and tree_mapping operations over them. This should allow unique number of groups per pixel. 

We can also set the variance to be large above those thresholds, but that will break some other summary stats (such as mean z-score)

**Comment by maxecharles** [comment]
Good luck with that lol

**Comment by maxecharles** [comment]
I've done the setting the variance large already and it was just more RAM intensive and didn't seem very necessary for the sources I was working with anyway so I sorta dropped it. Might become more necessary with quickly saturating lower group data.

**Comment by LouisDesdoigts** [comment]
Yeah our next set of CAL data blasts all the way up the ramp, so we will need to sort this out quite soon

**Comment by maxecharles** [comment]
expectation:
> Yeah our next set of CAL data blasts all the way up the ramp, so we will need to sort this out quite soon

reality: ~30k counts 🫠
